addons/test_mail/tests/test_mail_composer.py

```diff
@@ -0,0 +1,288 @@
+
+import base64
+
+from odoo.addons.mail.tests.common import mail_new_test_user
+from odoo.addons.test_mail.tests.common import TestMailCommon, TestRecipients
+from odoo.tests import tagged
+from odoo.tests.common import users, Form
+from odoo.tools import mute_logger
+
+@tagged('mail_composer')
+class TestMailComposer(TestMailCommon, TestRecipients):
+    """ Test Composer internals """
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestMailComposer, cls).setUpClass()
+        cls._init_mail_gateway()
+
+        cls.user_employee_2 = mail_new_test_user(
+            cls.env, login='employee2', groups='base.group_user',
+            notification_type='inbox', email='eglantine@example.com',
+            name='Eglantine Employee', signature='--\nEglantine')
+        cls.partner_employee_2 = cls.user_employee_2.partner_id
+
+        cls.test_record = cls.env['mail.test.ticket'].with_context(cls._test_context).create({
+            'name': 'TestRecord',
+            'customer_id': cls.partner_1.id,
+            'user_id': cls.user_employee_2.id,
+        })
+        cls.test_records = cls._create_records_for_batch('mail.test.ticket', 2)
+
+        cls.test_report = cls.env['ir.actions.report'].create({
+            'name': 'Test Report on mail test ticket',
+            'model': 'mail.test.ticket',
+            'report_type': 'qweb-pdf',
+            'report_name': 'test_mail.mail_test_ticket_test_template',
+        })
+
+        cls.test_from = '"John Doe" <john@example.com>'
+
+        cls.mail_server = cls.env['ir.mail_server'].create({
+            'name': 'Dummy Test Server',
+            'smtp_host': 'smtp.pizza.moc',
+            'smtp_port': 17,
+            'smtp_encryption': 'ssl',
+            'sequence': 666,
+        })
+
+        cls.template = cls._create_template('mail.test.ticket', {
+            'email_from': '${object.user_id.email_formatted | safe}',
+            'mail_server_id': cls.mail_server.id,
+        })
+
+    def _generate_attachments_data(self, count):
+        return [{
+            'name': '%02d.txt' % x,
+            'datas': base64.b64encode(b'Att%02d' % x),
+        } for x in range(count)]
+
+    def _get_context(self, records, add_web=True):
+        """ Helper to generate composer context. Will make tests a bit less
+        verbose.
+
+        :param add_web: add web context, generally making noise especially in
+          mass mail mode (active_id/ids both present in context)
+        """
+        base_context = {
+            'default_model': records._name,
+        }
+        if len(records) == 1:
+            base_context['default_composition_mode'] = 'comment'
+            base_context['default_res_id'] = records.id
+        else:
+            base_context['default_composition_mode'] = 'mass_mail'
+            base_context['active_ids'] = records.ids
+        if add_web:
+            base_context['active_model'] = records._name
+            base_context['active_id'] = records[0].id
```

Comment:
Should this be inside the `if len(records) == 1:` test?

Reply:
When testing various actions I often have both active_id and active_ids in context ... not sure. Anway I don't think it has a big impact as if composer behaves well with both of them, one of them is ok.